receiver/vmmetricsreceiver/vm_metrics_collector.go

```diff
@@ -113,104 +103,107 @@ func (vmc *VMMetricsCollector) StopCollection() {
 	close(vmc.done)
 }
 
-func (vmc *VMMetricsCollector) scrape(prevProcStat *procfs.ProcStat, prevStat *procfs.Stat) (*procfs.ProcStat, *procfs.Stat) {
+func (vmc *VMMetricsCollector) scrapeAndExport() {
+	ctx, span := trace.StartSpan(context.Background(), "VMMetricsCollector.scrapeAndExport")
+	defer span.End()
+
+	metrics := make([]*metricspb.Metric, 0, len(vmMetricDescriptors))
+	var errs []error
+
 	ms := &runtime.MemStats{}
 	runtime.ReadMemStats(ms)
-	ctx := context.Background()
-	stats.Record(
-		ctx,
-		mRuntimeAllocMem.M(int64(ms.Alloc)),
-		mRuntimeTotalAllocMem.M(int64(ms.TotalAlloc)),
-		mRuntimeSysMem.M(int64(ms.Sys)))
+	metrics = append(
+		metrics,
+		&metricspb.Metric{
+			MetricDescriptor: metricAllocMem,
+			Timeseries:       []*metricspb.TimeSeries{vmc.getInt64TimeSeries(ms.Alloc)},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricTotalAllocMem,
+			Timeseries:       []*metricspb.TimeSeries{vmc.getInt64TimeSeries(ms.TotalAlloc)},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricSysMem,
+			Timeseries:       []*metricspb.TimeSeries{vmc.getInt64TimeSeries(ms.Sys)},
+		},
+	)
 
 	pid := os.Getpid()
 	var proc procfs.Proc
 	var err error
-	if vmc.processFs == "" {
-		proc, err = vmc.fs.NewProc(pid)
-	} else {
-		proc, err = vmc.processFs.NewProc(pid)
-	}
+	proc, err = vmc.processFs.NewProc(pid)
```

Comment:
It seems that from `os.Getpid()` to here shouldn't happen on a timely basis, i.e. results could be saved and reused.

Reply:
Good suggestion, I moved it as part of `VMMetricsCollector` instead.